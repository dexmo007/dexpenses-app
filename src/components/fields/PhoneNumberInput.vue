<template>
  <v-text-field
    ref="input"
    v-bind="$attrs"
    v-on="$listeners"
    :value="model"
    @input="model = $event; $emit('input', $event)"
    :rules="[(v) => validatePhoneNumber(v) || 'Invalid phone number.']"
    :hint="showCountryPrefixHint ? 'It is recommended to add a country prefix, e.g. +49.' : null"
  >
    <template v-slot:append>
      <v-icon
        v-if="hasChanges()"
        @click="model = value"
      >clear</v-icon>
      <v-icon
        v-if="hasChanges() && valid"
        @click="$emit('save', model)"
      >save</v-icon>
    </template>
  </v-text-field>
</template>
<script>
export default {
  props: {
    value: String,
    countryPrefixHint: Boolean,
  },
  data() {
    return {
      model: this.value || null,
    };
  },
  computed: {
    valid() {
      try {
        return this.$refs.input.valid;
      } catch {
        return false;
      }
    },
    showCountryPrefixHint() {
      if (!this.countryPrefixHint) {
        return false;
      }
      if (!this.model) {
        return !this.hasCountryPrefix(this.value);
      }
      return !this.hasCountryPrefix(this.model);
    },
  },
  methods: {
    validatePhoneNumber(v) {
      return !v || !!v.match(/[\d\-()/\s]{5,}/);
    },
    hasCountryPrefix(v) {
      return v && v.match(/^(\+|00)\d\d/);
    },
    hasChanges() {
      if (this.model === null) {
        return false;
      }
      if (!this.value && !this.model) {
        return false;
      }
      return this.value !== this.model;
    },
    focus() {
      this.$refs.input.focus();
    },
  },
  watch: {
    value(v) {
      this.model = v || null;
    },
  },
};
</script>
