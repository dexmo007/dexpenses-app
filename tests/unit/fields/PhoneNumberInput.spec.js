import { createLocalVue, mount } from '@vue/test-utils';
import Vuetify from 'vuetify';
import PhoneNumberInput from '@/components/fields/PhoneNumberInput.vue';

const localVue = createLocalVue();

describe('PhoneNumberInput', () => {
  let vuetify;
  beforeEach(() => {
    vuetify = new Vuetify({
      icons: {
        iconfont: 'md',
      },
    });
  });

  it('should render correctly', () => {
    const wrapper = mount(PhoneNumberInput, {
      propsData: {
        value: '+491234567890',
      },
      localVue,
      vuetify,
    });
    expect(wrapper.element).toMatchSnapshot();

    expect(wrapper.find('input').element.value).toBe('+491234567890');
  });

  it('should show country code prefix hint', async () => {
    const wrapper = mount(PhoneNumberInput, {
      propsData: {
        value: '01234567890',
        countryPrefixHint: true,
      },
      localVue,
      vuetify,
    });
    wrapper.find('input').element.focus();
    expect(wrapper.element).toMatchSnapshot(); // show hint
    wrapper.find('input').setValue('+491234567890');
    expect(wrapper.element).toMatchSnapshot(); // hide hint
  });

  it('should show save and restore button if changes are made', () => {
    const wrapper = mount(PhoneNumberInput, {
      propsData: {
        value: '+491234567890',
      },
      localVue,
      vuetify,
    });
    expect(wrapper.element).toMatchSnapshot(); // no buttons
    wrapper.find('input').setValue('+491234567891');
    expect(wrapper.element).toMatchSnapshot(); // with buttons
  });

  it('should show save and restore button if input is cleared', () => {
    const wrapper = mount(PhoneNumberInput, {
      propsData: {
        value: '+491234567890',
      },
      localVue,
      vuetify,
    });
    expect(wrapper.element).toMatchSnapshot(); // no buttons
    wrapper.find('input').setValue('');
    expect(wrapper.element).toMatchSnapshot(); // with buttons
  });

  it('should show error and hide save button', async () => {
    const wrapper = mount(PhoneNumberInput, {
      propsData: {
        value: '+491234567890',
      },
      localVue,
      vuetify,
    });
    wrapper.find('input').setValue('+4912');
    await wrapper.vm.$nextTick(); // wait for validation changes
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should restore saved value', () => {
    const wrapper = mount(PhoneNumberInput, {
      propsData: {
        value: '+491234567890',
      },
      localVue,
      vuetify,
    });
    wrapper.find('input').setValue('+491234567891');
    const clear = wrapper.findAll('i').wrappers.find(w => w.text() === 'clear');
    expect(clear).toBeDefined();
    clear.trigger('click');
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.vm.model).toBe('+491234567890');
  });

  it('should save value', () => {
    const saveCallback = jest.fn();
    const wrapper = mount(PhoneNumberInput, {
      propsData: {
        value: '+491234567890',
      },
      listeners: {
        save: saveCallback,
      },
      localVue,
      vuetify,
    });
    wrapper.find('input').setValue('+491234567891');
    const save = wrapper.findAll('i').wrappers.find(w => w.text() === 'save');
    expect(save).toBeDefined();
    save.trigger('click');
    expect(saveCallback).toHaveBeenCalled();
    wrapper.setProps({ value: '+491234567891' }); // assume persistent value
    expect(wrapper.element).toMatchSnapshot(); // clean state after save
  });
});
