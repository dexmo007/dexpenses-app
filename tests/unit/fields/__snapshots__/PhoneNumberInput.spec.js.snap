// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneNumberInput should render correctly 1`] = `
<div
  class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field"
>
  <div
    class="v-input__control"
  >
    <div
      class="v-input__slot"
    >
      <div
        class="v-text-field__slot"
      >
        <input
          id="input-3"
          type="text"
        />
      </div>
      <div
        class="v-input__append-inner"
      >
        <!---->
         
        <!---->
      </div>
    </div>
    <div
      class="v-text-field__details"
    >
      <div
        class="v-messages theme--light"
      >
        <span
          class="v-messages__wrapper"
          name="message-transition"
          tag="div"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`PhoneNumberInput should restore saved value 1`] = `
<div
  class="v-input v-input--is-label-active v-input--is-dirty v-input--is-focused theme--light v-text-field primary--text"
>
  <div
    class="v-input__control"
  >
    <div
      class="v-input__slot"
    >
      <div
        class="v-text-field__slot"
      >
        <input
          id="input-41"
          type="text"
        />
      </div>
      <div
        class="v-input__append-inner"
      >
        <!---->
         
        <!---->
      </div>
    </div>
    <div
      class="v-text-field__details"
    >
      <div
        class="v-messages theme--light primary--text"
      >
        <span
          class="v-messages__wrapper"
          name="message-transition"
          tag="div"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`PhoneNumberInput should save value 1`] = `
<div
  class="v-input v-input--is-label-active v-input--is-dirty v-input--is-focused theme--light v-text-field primary--text"
>
  <div
    class="v-input__control"
  >
    <div
      class="v-input__slot"
    >
      <div
        class="v-text-field__slot"
      >
        <input
          id="input-49"
          type="text"
        />
      </div>
      <div
        class="v-input__append-inner"
      >
        <!---->
         
        <!---->
      </div>
    </div>
    <div
      class="v-text-field__details"
    >
      <div
        class="v-messages theme--light primary--text"
      >
        <span
          class="v-messages__wrapper"
          name="message-transition"
          tag="div"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`PhoneNumberInput should show country code prefix hint 1`] = `
<div
  class="v-input v-input--is-label-active v-input--is-dirty v-input--is-focused theme--light v-text-field primary--text"
>
  <div
    class="v-input__control"
  >
    <div
      class="v-input__slot"
    >
      <div
        class="v-text-field__slot"
      >
        <input
          id="input-9"
          type="text"
        />
      </div>
      <div
        class="v-input__append-inner"
      >
        <!---->
         
        <!---->
      </div>
    </div>
    <div
      class="v-text-field__details"
    >
      <div
        class="v-messages theme--light"
      >
        <span
          class="v-messages__wrapper"
          name="message-transition"
          tag="div"
        >
          <div
            class="v-messages__message"
          >
            It is recommended to add a country prefix, e.g. +49.
          </div>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`PhoneNumberInput should show country code prefix hint 2`] = `
<div
  class="v-input v-input--is-label-active v-input--is-dirty v-input--is-focused theme--light v-text-field primary--text"
>
  <div
    class="v-input__control"
  >
    <div
      class="v-input__slot"
    >
      <div
        class="v-text-field__slot"
      >
        <input
          id="input-9"
          type="text"
        />
      </div>
      <div
        class="v-input__append-inner"
      >
        <i
          class="v-icon notranslate v-icon--link material-icons theme--light"
          role="button"
        >
          clear
        </i>
         
        <i
          class="v-icon notranslate v-icon--link material-icons theme--light"
          role="button"
        >
          save
        </i>
      </div>
    </div>
    <div
      class="v-text-field__details"
    >
      <div
        class="v-messages theme--light primary--text"
      >
        <span
          class="v-messages__wrapper"
          name="message-transition"
          tag="div"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`PhoneNumberInput should show error and hide save button 1`] = `
<div
  class="v-input v-input--has-state v-input--is-label-active v-input--is-dirty theme--light v-text-field error--text"
>
  <div
    class="v-input__control"
  >
    <div
      class="v-input__slot"
    >
      <div
        class="v-text-field__slot"
      >
        <input
          id="input-33"
          type="text"
        />
      </div>
      <div
        class="v-input__append-inner"
      >
        <i
          class="v-icon notranslate v-icon--link material-icons theme--light"
          role="button"
        >
          clear
        </i>
         
        <!---->
      </div>
    </div>
    <div
      class="v-text-field__details"
    >
      <div
        class="v-messages theme--light error--text"
      >
        <span
          class="v-messages__wrapper"
          name="message-transition"
          tag="div"
        >
          <div
            class="v-messages__message"
          >
            Invalid phone number.
          </div>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`PhoneNumberInput should show save and restore button if changes are made 1`] = `
<div
  class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field"
>
  <div
    class="v-input__control"
  >
    <div
      class="v-input__slot"
    >
      <div
        class="v-text-field__slot"
      >
        <input
          id="input-17"
          type="text"
        />
      </div>
      <div
        class="v-input__append-inner"
      >
        <!---->
         
        <!---->
      </div>
    </div>
    <div
      class="v-text-field__details"
    >
      <div
        class="v-messages theme--light"
      >
        <span
          class="v-messages__wrapper"
          name="message-transition"
          tag="div"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`PhoneNumberInput should show save and restore button if changes are made 2`] = `
<div
  class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field"
>
  <div
    class="v-input__control"
  >
    <div
      class="v-input__slot"
    >
      <div
        class="v-text-field__slot"
      >
        <input
          id="input-17"
          type="text"
        />
      </div>
      <div
        class="v-input__append-inner"
      >
        <i
          class="v-icon notranslate v-icon--link material-icons theme--light"
          role="button"
        >
          clear
        </i>
         
        <i
          class="v-icon notranslate v-icon--link material-icons theme--light"
          role="button"
        >
          save
        </i>
      </div>
    </div>
    <div
      class="v-text-field__details"
    >
      <div
        class="v-messages theme--light"
      >
        <span
          class="v-messages__wrapper"
          name="message-transition"
          tag="div"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`PhoneNumberInput should show save and restore button if input is cleared 1`] = `
<div
  class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field"
>
  <div
    class="v-input__control"
  >
    <div
      class="v-input__slot"
    >
      <div
        class="v-text-field__slot"
      >
        <input
          id="input-25"
          type="text"
        />
      </div>
      <div
        class="v-input__append-inner"
      >
        <!---->
         
        <!---->
      </div>
    </div>
    <div
      class="v-text-field__details"
    >
      <div
        class="v-messages theme--light"
      >
        <span
          class="v-messages__wrapper"
          name="message-transition"
          tag="div"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`PhoneNumberInput should show save and restore button if input is cleared 2`] = `
<div
  class="v-input theme--light v-text-field"
>
  <div
    class="v-input__control"
  >
    <div
      class="v-input__slot"
    >
      <div
        class="v-text-field__slot"
      >
        <input
          id="input-25"
          type="text"
        />
      </div>
      <div
        class="v-input__append-inner"
      >
        <i
          class="v-icon notranslate v-icon--link material-icons theme--light"
          role="button"
        >
          clear
        </i>
         
        <i
          class="v-icon notranslate v-icon--link material-icons theme--light"
          role="button"
        >
          save
        </i>
      </div>
    </div>
    <div
      class="v-text-field__details"
    >
      <div
        class="v-messages theme--light"
      >
        <span
          class="v-messages__wrapper"
          name="message-transition"
          tag="div"
        />
      </div>
    </div>
  </div>
</div>
`;
